stellarphot/settings/tests/test_views.py

```diff
@@ -0,0 +1,68 @@
+from stellarphot.settings import Camera, ui_generator
+from stellarphot.settings.tests.test_models import TEST_CAMERA_VALUES
+
+
+class TestUiGenerator:
+    def test_camera(self):
+        ui = ui_generator(Camera)
+        # The description should be the beginning of the docstring
+        assert Camera.__doc__.strip().startswith(ui.description.split()[0])
+
+        # We always want to show nullable fields
+        assert ui.show_null
+
+        # Which means we don't need the button to show/hide them
+        assert ui.bn_shownull.layout.display == "none"
+
+        # For now we do not show the validaation output because it is painfully
+        # verbose and not very helpful.
+        assert not ui.show_validation
+
+        # We always display nested models
+        assert ui.open_nested
+
+        # Finally, we don't want the widget to update continuously because it
+        # will overwrite the value the user entered.
+        for widget in ui.di_widgets.values():
+            if hasattr(widget, "continuous_update"):
+                assert not widget.continuous_update
+
+    def test_disabled_state_save_revert_button(self):
+        # We want the save and revert buttons to be enabled only when
+        # 1. The user has made a change, AND
+        # 2. The value in the widget is a valid pydantic model
+
+        ui = ui_generator(Camera)
+        # The save button should be disabled
+        assert ui.savebuttonbar.bn_save.disabled
+        assert ui.savebuttonbar.bn_revert.disabled
+
+        # Set one field to a valid value....
+        ui.value["name"] = "test"
+
+        # ...and the save button should still be disabled
+        assert ui.savebuttonbar.bn_save.disabled
+        assert ui.savebuttonbar.bn_revert.disabled
+
+        # Set a valid value
+        ui.value = TEST_CAMERA_VALUES
+
+        # So it turns out that the validation stuff only updates when changes are made
+        # in the UI rather than programmatically. Since we know we've set a valid value,
+        # and that we've made changes we just manually set the relevant values.
```

Comment:
This is rather sad, forcing the test to work, but I suppose it checks that changing `ui.savebuttonbar.unsaved_changes` and ` ui.is_valid.value` have the intended effects on the save button.

Reply:
Yeah, this was not my first choice. I tried calling the handlers that I thought `ipyautoui` was invoking, but i) they were private (started with `_`) and ii) didn't do anything.